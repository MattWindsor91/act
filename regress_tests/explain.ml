(* This file is part of 'act'.

   Copyright (c) 2018, 2019 by Matt Windsor

   Permission is hereby granted, free of charge, to any person obtaining a
   copy of this software and associated documentation files (the
   "Software"), to deal in the Software without restriction, including
   without limitation the rights to use, copy, modify, merge, publish,
   distribute, sublicense, and/or sell copies of the Software, and to permit
   persons to whom the Software is furnished to do so, subject to the
   following conditions:

   The above copyright notice and this permission notice shall be included
   in all copies or substantial portions of the Software.

   THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
   OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
   MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN
   NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
   DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR
   OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE
   USE OR OTHER DEALINGS IN THE SOFTWARE. *)

(** Tests the output of the 'act asm explain' command. *)

open Core
module Ac = Act_common

let run (dir : Fpath.t) : unit Or_error.t =
  let arch = Ac.Id.of_string "x86.att" in
  Or_error.Let_syntax.(
    let%bind (module L) =
      Toplevel.Language_support.asm_runner_from_arch arch
    in
    let module Exp = Act_asm.Explainer.Make (L) in
    Common.regress_run_asm_many
      (module Exp.Filter)
      "Explainer" Act_sanitiser.Pass_group.explain dir
      ~config_fn:(fun aux -> Act_asm.Explainer.Config.make ~aux ()) 
  )

let command =
  Common.make_regress_command ~summary:"runs explainer regressions" run

let () = Command.run command
