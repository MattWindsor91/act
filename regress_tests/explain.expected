# Explainer tests

## clang_macho_o3_xchg.s

```
-- program 0
_P0:  # <-- label {unused label, program boundary}
pushl %EBP # <-- stack {manipulates stack}
movl %ESP, %EBP # <-- move {manipulates stack}
movl $2, %EAX # <-- move
xchgl %EAX, _x # <-- RMW
movl $1, %EAX # <-- move
xchgl %EAX, _y # <-- RMW
popl %EBP # <-- stack {manipulates stack}
retl # <-- return
-- program 1
_P1:  # <-- label {unused label, program boundary}
pushl %EBP # <-- stack {manipulates stack}
movl %ESP, %EBP # <-- move {manipulates stack}
movl $2, %EAX # <-- move
xchgl %EAX, _y # <-- RMW
movl $1, %EAX # <-- move
xchgl %EAX, _x # <-- RMW
popl %EBP # <-- stack {manipulates stack}
retl # <-- return
```
## gcc_elf_o0_stackchk.s

```
-- program 0
P0:  # <-- label {unused label, program boundary}
pushl %EBP # <-- stack {manipulates stack}
movl %ESP, %EBP # <-- move {manipulates stack}
subl $24, %ESP # <-- arith {manipulates stack}
movl %GS:20, %EAX # <-- move
movl %EAX, -12(%EBP) # <-- move
xorl %EAX, %EAX # <-- logical
movl $2, -20(%EBP) # <-- move
movl -20(%EBP), %EAX # <-- move
movl %EAX, y # <-- move
mfence # <-- fence
movl x, %EAX # <-- move
movl %EAX, -16(%EBP) # <-- move
movl -16(%EBP), %EAX # <-- move
movl %EAX, t0r0 # <-- move
movl -12(%EBP), %EAX # <-- move
xorl %GS:20, %EAX # <-- logical
je .L2 # <-- jump
call __stack_chk_fail # <-- call
.L2:  # <-- label
leave # <-- call
ret # <-- return
-- program 1
P1:  # <-- label {unused label, program boundary}
pushl %EBP # <-- stack {manipulates stack}
movl %ESP, %EBP # <-- move {manipulates stack}
subl $24, %ESP # <-- arith {manipulates stack}
movl %GS:20, %EAX # <-- move
movl %EAX, -12(%EBP) # <-- move
xorl %EAX, %EAX # <-- logical
movl $1, -20(%EBP) # <-- move
movl -20(%EBP), %EAX # <-- move
movl %EAX, x # <-- move
mfence # <-- fence
movl $1, -16(%EBP) # <-- move
movl -16(%EBP), %EAX # <-- move
movl %EAX, y # <-- move
mfence # <-- fence
movl -12(%EBP), %EAX # <-- move
xorl %GS:20, %EAX # <-- logical
je .L4 # <-- jump
call __stack_chk_fail # <-- call
.L4:  # <-- label
leave # <-- call
ret # <-- return
```
## gcc_macho_o0_deref_cmpxchg.s

```
-- program 0
_P0:  # <-- label {unused label, program boundary}
pushl %EBP # <-- stack {manipulates stack}
movl %ESP, %EBP # <-- move {manipulates stack}
subl $16, %ESP # <-- arith {manipulates stack}
movl $0, _t0r0 # <-- move
movl $_x, -4(%EBP) # <-- move
movl $2, -8(%EBP) # <-- move
movl -8(%EBP), %EAX # <-- move
movl %EAX, %EDX # <-- move
movl -4(%EBP), %ECX # <-- move
movl _t0r0, %EAX # <-- move
lock cmpxchgl %EDX, (%ECX) # <-- RMW
movl %EAX, %EDX # <-- move
sete %AL # <-- ??
testb %AL, %AL # <-- ??
jne L3 # <-- jump
movl %EDX, _t0r0 # <-- move
L3:  # <-- label
leave # <-- call
ret # <-- return
-- program 1
_P1:  # <-- label {unused label, program boundary}
pushl %EBP # <-- stack {manipulates stack}
movl %ESP, %EBP # <-- move {manipulates stack}
subl $16, %ESP # <-- arith {manipulates stack}
movl $_x, -4(%EBP) # <-- move
movl $1, -8(%EBP) # <-- move
movl -8(%EBP), %EAX # <-- move
movl %EAX, %EDX # <-- move
movl -4(%EBP), %EAX # <-- move
movl %EDX, (%EAX) # <-- move
leave # <-- call
ret # <-- return
```
## gcc_macho_o0_deref_write.s

```
-- program 0
_P0:  # <-- label {unused label, program boundary}
pushl %EBP # <-- stack {manipulates stack}
movl %ESP, %EBP # <-- move {manipulates stack}
subl $16, %ESP # <-- arith {manipulates stack}
movl $_y, -4(%EBP) # <-- move
movl -4(%EBP), %EAX # <-- move
movl (%EAX), %EAX # <-- move
movl %EAX, -8(%EBP) # <-- move
movl -8(%EBP), %EAX # <-- move
movl %EAX, _t0r0 # <-- move
movl _t0r0, %EAX # <-- move
cmpl $1, %EAX # <-- compare
jne L3 # <-- jump
movl $1, _x # <-- move
L3:  # <-- label
leave # <-- call
ret # <-- return
-- program 1
_P1:  # <-- label {unused label, program boundary}
pushl %EBP # <-- stack {manipulates stack}
movl %ESP, %EBP # <-- move {manipulates stack}
subl $16, %ESP # <-- arith {manipulates stack}
movl _x, %EAX # <-- move
movl %EAX, _t1r0 # <-- move
movl _t1r0, %EAX # <-- move
cmpl $1, %EAX # <-- compare
jne L6 # <-- jump
movl $_y, -4(%EBP) # <-- move
movl $1, -8(%EBP) # <-- move
movl -8(%EBP), %EAX # <-- move
movl %EAX, %EDX # <-- move
movl -4(%EBP), %EAX # <-- move
movl %EDX, (%EAX) # <-- move
L6:  # <-- label
leave # <-- call
ret # <-- return
```
## gcc_macho_o0_immediate_heap.s

```
-- program 0
_P0:  # <-- label {unused label, program boundary}
pushl %EBP # <-- stack {manipulates stack}
movl %ESP, %EBP # <-- move {manipulates stack}
subl $16, %ESP # <-- arith {manipulates stack}
movl $_x, -4(%EBP) # <-- move
movl -4(%EBP), %EAX # <-- move
movl (%EAX), %EAX # <-- move
movl %EAX, -12(%EBP) # <-- move
movl -12(%EBP), %EAX # <-- move
movl %EAX, _t0r0 # <-- move
movl $_y, -8(%EBP) # <-- move
movl $1, -16(%EBP) # <-- move
movl -16(%EBP), %EAX # <-- move
movl %EAX, %EDX # <-- move
movl -8(%EBP), %EAX # <-- move
movl %EDX, (%EAX) # <-- move
leave # <-- call
ret # <-- return
-- program 1
_P1:  # <-- label {unused label, program boundary}
pushl %EBP # <-- stack {manipulates stack}
movl %ESP, %EBP # <-- move {manipulates stack}
subl $16, %ESP # <-- arith {manipulates stack}
movl $_y, -4(%EBP) # <-- move
movl $2, -12(%EBP) # <-- move
movl -12(%EBP), %EAX # <-- move
movl %EAX, %EDX # <-- move
movl -4(%EBP), %EAX # <-- move
movl %EDX, (%EAX) # <-- move
movl $_x, -8(%EBP) # <-- move
movl $1, -16(%EBP) # <-- move
movl -16(%EBP), %EAX # <-- move
movl %EAX, %EDX # <-- move
movl -8(%EBP), %EAX # <-- move
movl %EDX, (%EAX) # <-- move
leave # <-- call
ret # <-- return
```
## gcc_macho_o0_moves.s

```
-- program 0
_P0:  # <-- label {unused label, program boundary}
pushl %EBP # <-- stack {manipulates stack}
movl %ESP, %EBP # <-- move {manipulates stack}
subl $16, %ESP # <-- arith {manipulates stack}
movl _y, %EAX # <-- move
movl %EAX, -4(%EBP) # <-- move
movl -4(%EBP), %EAX # <-- move
movl %EAX, _t0r0 # <-- move
movl _t0r0, %EAX # <-- move
cmpl $1, %EAX # <-- compare
jne L3 # <-- jump
movl $1, _x # <-- move
L3:  # <-- label
leave # <-- call
ret # <-- return
-- program 1
_P1:  # <-- label {unused label, program boundary}
pushl %EBP # <-- stack {manipulates stack}
movl %ESP, %EBP # <-- move {manipulates stack}
subl $16, %ESP # <-- arith {manipulates stack}
movl _x, %EAX # <-- move
movl %EAX, _t1r0 # <-- move
movl _t1r0, %EAX # <-- move
cmpl $1, %EAX # <-- compare
jne L6 # <-- jump
movl $1, -4(%EBP) # <-- move
movl -4(%EBP), %EAX # <-- move
movl %EAX, _y # <-- move
mfence # <-- fence
L6:  # <-- label
leave # <-- call
ret # <-- return
```

Ran 6 test(s).
